File: src/main/java/com/raychenon/leetcode/BackspaceStringCompare.java
package com.raychenon.leetcode;

import java.util.Stack;

/**
 * User: raychenon
 * Date: 9/4/2020
 * https://leetcode.com/problems/backspace-string-compare
 */
public class BackspaceStringCompare {

    public boolean backspaceCompare(String S, String T) {
        String bs = backspaced(S);
        String bt = backspaced(T);
        System.out.println("bs = " + bs + ", bt = " + bt);
        return bt.equals(bs);
    }

    private String backspaced(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c != '#') {
                stack.add(c);
            } else if (!stack.isEmpty()) {
                stack.pop();
            }
        }

        return stack.toString();
    }

}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/BackspaceStringCompare.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 9/4/2020
 * https://leetcode.com/problems/backspace-string-compare
 */
object BackspaceStringCompare {

    fun backspaceCompare(S: String, T: String): Boolean {
        return backspaced(S).equals(backspaced(T))
    }

    fun backspaced(s: String): String {
        var str = StringBuilder()
        for (i in 0 until s.length) {
            val c = s.get(i)
            if (c == '#') {
                val len = str.length
                if (len > 0) str.deleteCharAt(len - 1)
            } else {
                str.append(c)
            }
        }
        return str.toString()
    }

}